File: CodeMaze-master/src/Ladder_Algorithm/3-BinaryTree_DivideConquer/93_BalancedBinaryTree/solution2.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */

class ResultType {
public:
    bool isBalanced;
    int maxDepth;
    ResultType(bool isBalanced, int maxDepth) {
        this->isBalanced = isBalanced;
        this->maxDepth = maxDepth;
    }
};

class Solution {
public:
    /**
     * @param root: The root of binary tree.
     * @return: True if this Binary tree is Balanced, or false.
     */
    bool isBalanced(TreeNode *root) {
        // write your code here
        return helper(root)->isBalanced;
    }

private:
    ResultType* helper(TreeNode *node) {
        if (node == NULL) {
            return new ResultType(true, 0);
        }
        ResultType *left = helper(node->left);
        ResultType *right = helper(node->right);

        if (!left->isBalanced || !right->isBalanced) {
            return new ResultType(false, INT_MIN);
        }

        if (abs(left->maxDepth - right->maxDepth) > 1) {
            return new ResultType(false, INT_MIN);
        }

        return new ResultType(true, max(left->maxDepth, right->maxDepth) + 1);

        /* or
        if (left->isBalanced && right->isBalanced && abs(left->maxDepth - right->maxDepth) <= 1) {
            int h = max(left->maxDepth, right->maxDepth) + 1;
            return new ResultType(true, h);
        } else {
            return new ResultType(false, 0);
        }
        */

    }

};

File: CodeMaze-master/src/best_time_to_buy_and_sell_stock/solution2.java
class Solution {
    public int maxProfit(int[] prices) {
        int max;
        int l = prices.length;
        if (l == 0 || l == 1) {
            return 0;
        }
        int[] d = new int[l];
        d[l-2] = prices[l-1] - prices[l-2];
        max = d[l-2];
        for (int i = l - 3; i >= 0; --i) {
            d[i] = ((prices[i+1] > d[i+1] + prices[i+1]) ? prices[i+1] : d[i+1] + prices[i+1]) - prices[i];
            if (max < d[i]) {
                max = d[i];
            }
        }
        if (max < 0) return 0;
        return max;
    }
}